tasmota/tasmota_xsns_sensor/xsns_14_sht3x.ino

```diff
@@ -104,29 +106,32 @@ bool Sht3xRead(uint32_t sensor, float &t, float &h) {
   if ((Sht3xComputeCrc(&data[0], 2) != data[2]) || (Sht3xComputeCrc(&data[3], 2) != data[5])) {
     return false;
   }
-  t = ((float)(((data[0] << 8) | data[1]) * 175) / 65535.0) - 45.0;
+  float t;
+  float h;
+  t = ((((data[0] << 8) | data[1]) * 175) / 65535.0) - 45.0;
```

Comment:
You can save 48 bytes by changing all `double` literals to `float` literals. I.e. `65536.0f` instead of `65536.0`...

When you don't add the `f` suffix, all calcucations are done in `double` precision instead of `float` single precision, which generally brings unwanted libraries and takes more time and space. For temp/humi, single precision is more than enough

Reply:
Is there anything to be said against integer arithmetic by using tenths or hundredths as the basis everywhere?

Reply:
That would be even better. For ex:

```C
uint16_t changeUIntScale(uint16_t inum, uint16_t ifrom_min, uint16_t ifrom_max,
                                       uint16_t ito_min, uint16_t ito_max)
```

is a easy way to convert unsigned 16 bits from range a..b to c..d.

So if I'm not mistaken, it should be something like:

```C
t = changeUIntScale((data[0] << 8) | data[1], 0, 65536, -45, 130);
```

Reply:
Hmmm, the example above would lose precision. The alternative would be to have `int16_t t_10` as the temperature in 1/10th of Celsius, and:

```C
t = changeIntScale((data[0] << 8) | data[1], 0, 65535, -450, 1300);
```

1/100th of Celsius would overflow the `int16_t` size.

Reply:
Good call. While the resolution is 1/100th, the accuracy of these sensors is at best 0.1 °C (SHT45) and typically 0.2 °C (SHT40). Do we sacrifice resolution for code size? What are your thoughts?

Reply:
I generally tend to prefer integers, so yes in 1/10th resolution should be enough.

The way I do in Zigbee to print the value is:
```C
              char buf[12];
              dtostrf(t_100 / 100.0f, 2, 1, buf);
```

But `float` is fine as well

Reply:
> 1/100th of Celsius would overflow the `int16_t` size.

One decimal place should be sufficient.

Reply:
Great. Who creates a new PR? I can do it tomorrow the earliest.

Reply:
@s-hadinger @eku 
I have another question regarding this. I just played around with different data types and I noticed 
that int16_t allocates more memory than int32_t since a conversion happens.
I think this is briefly touched in https://tasmota.github.io/docs/Sensor-API/#uint8_t-or-uint32_t

Wouldn't it be recommended to change, apart from `t_10`->int32_t and `h_10`->int32_t, e.g., `sht3x_count` and `sht3x_addressed`, and `data` to **u**int32_t as well and only preserve the `uint8_t`s in the struct?

And what is actually the best practice regarding explicity defining signed/unsigned integers over a simple `int`?

Reply:
Some of my best? practices:
* Type int is always signed. Prefer unsigned if you don’t need negative values.
* Don't use int or unsigned if you need more than 16bit (for portability). Use int32_t and friends for specific size requirements.
* On tasmota prefer smallest data over smaller flash over most performant. RAM is the most constrained resource.

Reply:
While 1 decimal may be "sufficient" on a temperature, I'd not support removing precision from what the sensor provides. While 2 decimals will be more than the accuracy you can expect from a sensor, the 2nd decimal is pretty good for trends, meaning that a bot of temperatures does not get so jagged, and you can also spot small trends if you're after fine-grained control. Also even if the sensor may only have a accuracy of 0.5.

Reply:
Thank you @joba-1 and @sfromis for the clarification.
With reference to the example here, we can probably conclude that there is no better way of calculating the temperature without sacrificing precision and this is as good as it gets:
`float t = ((((data0 << 8) | data1) * 175) / 65535.0f) - 45;` 

In terms of relative humidity, I think there is some potential without sacrificing any physically meaningful precision:

`int32_t h_100 = ((((data[3] << 8) | data[4]) * 12500) >> 16) - 600;` and
`int32_t h_100 = ((((data[3] << 8) | data[4]) * 10000) >> 16);`
saves 32 bytes over 
`float h = (((data[3] << 8) | data[4]) * 100) / 65535.0f;` and
`float h  = ((((data[3] << 8) | data[4]) * 125) / 65535.0f) - 6;`

Agreed?

Reply:
Yes